Fabric Networking API V1

i509VCB opened the pull request:
This is partially based of off and supersedes #605. This pull request is based off the `feature/cut` branch on the fablabs drafting repository. This pull request was primarily written by @liach but has been adapted into a new pull request on the FabLabs fabric api fork to continue the pull request.

This pull request migrates the networking api to v1 and deprecates the old api (minus player streams which need to be migrated by a different pull request). The old networking api is re-implemented using the new networking api in a backwards compatible way.

Registration of custom packet handlers is now done through the client or server networking's play receiver. The interface used to read the packet and apply (originally PacketConsumer) is now much more ergonomic, as it exposes the client and serve respectively along with the network handler. This effectively resolves #345.

This pull request also adds the ability for mods to send and handle login query messages. This allows the client and server to communicate much earlier and delay login for the client/server until tasks have been completed on either side (async tasks included).

Events related to the connection/disconnection of the client or server were added to the respective `Client/ServerConnectionEvents`. This effectively resolves #247

Also resolves #187, and prepares the groundwork for a system described in #164.

i509VCB committed:
Deprecate v0 and implement using v1

i509VCB committed:
Networking api v1




Some final docs?


Licenses and testmod


Fix a bunch o imports and make things work for v1 (v0 is bork)


Make the testmod pass checkstyle and work


Docs for v1

i509VCB committed:
Drop files down one package due to package check error

i509VCB committed:
Fix issue with channel registration, add another testmod

i509VCB committed:
jaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaavadoc

i509VCB committed:
Make javadoc use `code`, move impl interface to package access

i509VCB committed:
this things

i509VCB committed:
Rename a few internal methods

i509VCB committed:
Mark all client side stuff client only, move client mixins

RDIL commented on fabric-networking-api-v1/src/main/java/net/fabricmc/fabric/api/networking/v1/ClientConnectionEvents.java:
```diff
@@ -0,0 +1,69 @@
+	 *
+	 * <p>No packets should be sent when this event is invoked.
+	 */
+	public static final Event<PlayDisconnected> PLAY_DISCONNECTED = EventFactory.createArrayBacked(PlayDisconnected.class, callbacks -> (handler, client) -> {
```
This might just be my computer but it looks like this indent is different then below?

i509VCB replied:
I think GH is funky with tabs sometimes

Vaerian commented on fabric-networking-api-v1/src/main/java/net/fabricmc/fabric/api/networking/v1/ClientChannelEvents.java:
```diff
@@ -0,0 +1,71 @@
+	 * An event for the client play network handler receiving an update indicating the connected server's ability to receive packets in certain channels.
+	 *
+	 * @see PlayPacketSender#hasChannel(Identifier)
+	 */
```
This could be further clarified. At first glance understanding what this does is a little difficult. Like when exactly does this trigger.

Vaerian commented on fabric-networking-api-v1/src/main/java/net/fabricmc/fabric/api/networking/v1/ClientChannelEvents.java:
```diff
@@ -0,0 +1,71 @@
+	 * An event for the client play network handler receiving an update indicating the connected server's lack of ability to receive packets in certain channels.
+	 *
+	 * @see PlayPacketSender#hasChannel(Identifier)
+	 */
```
Same here

Vaerian commented on fabric-networking-api-v1/src/main/java/net/fabricmc/fabric/api/networking/v1/ClientNetworking.java:
```diff
@@ -0,0 +1,151 @@
+	/**
+	 * Returns the packet receiver for channel handler registration on client play network handlers, receiving {@link net.minecraft.network.packet.s2c.play.CustomPayloadS2CPacket server to client custom payload packets}.
+	 */
+	public static ChannelHandlerRegistry<PlayChannelHandler> getPlayReceiver() {
```
I might change this to `getPlayReceivers` to clarify that this isn't returning one specific play receiver but a registry of play receivers

Vaerian commented on fabric-networking-api-v1/src/main/java/net/fabricmc/fabric/api/networking/v1/ClientNetworking.java:
```diff
@@ -0,0 +1,151 @@
+	/**
+	 * Returns the packet receiver for channel handler registration on client login network handlers, receiving {@link net.minecraft.network.packet.s2c.login.LoginQueryRequestS2CPacket login query request packets}.
+	 */
+	public static ChannelHandlerRegistry<LoginChannelHandler> getLoginReceiver() {
```
Same here `getLoginReceivers`

Vaerian commented on fabric-networking-api-v1/src/main/java/net/fabricmc/fabric/api/networking/v1/ClientNetworking.java:
```diff
@@ -0,0 +1,151 @@
+	 * @param channel the channel of the packet
+	 * @param buf the payload of the packet
+	 */
+	public static void send(Identifier channel, PacketByteBuf buf) {
```
Doesn't this have the opportunity to throw an exception? If so, that should be mentioned in the javadoc.

Vaerian replied:
Also maybe there should be a method similar to `getPlaySender(ClientPlayNetworkHandler handler)` and `getPlaySender() throws IllegalStateException` that is `send(ClientPlayNetworkHandler handler, Identifier channel, PacketByteBuf buf)`

Vaerian commented on fabric-networking-api-v1/src/main/java/net/fabricmc/fabric/api/networking/v1/ClientNetworking.java:
```diff
@@ -0,0 +1,151 @@
+		 * <p>This method is executed on {@linkplain io.netty.channel.EventLoop netty's event loops}.
+		 * Modification to the game should be {@linkplain net.minecraft.util.thread.ThreadExecutor#submit(Runnable) scheduled} using the provided Minecraft client instance.
+		 *
+		 * <p>An example usage of this is to display a
```
Where did the rest of this sentence go. I'm dying to know what it says

Vaerian commented on fabric-networking-api-v1/src/main/java/net/fabricmc/fabric/api/networking/v1/PlayPacketSender.java:
```diff
@@ -0,0 +1,29 @@
+ * Supports sending packets to channels in the play network handlers.
+ *
+ * <p>Compared to a simple packet sender, the play packet sender is informed if its connected recipient may {@link #hasChannel(Identifier) accept packets in certain channels}.
+ * <!--TODO: Implement use of system property-->
```
Just highlighting this as something that is on the to-do list for others looking at this review to get a synopsis

Vaerian commented on fabric-networking-api-v1/src/main/java/net/fabricmc/fabric/api/networking/v1/ServerChannelEvents.java:
```diff
@@ -0,0 +1,66 @@
+	 * An event for the server play network handler receiving an update indicating the connected client's lack of ability to receive packets in certain channels.
+	 *
+	 * @see PlayPacketSender#hasChannel(Identifier)
+	 */
```
Similar to the comment I made on the client side, more specifics about when this occurs would be good in the javadoc

Vaerian commented on fabric-networking-api-v1/src/main/java/net/fabricmc/fabric/api/networking/v1/ServerChannelEvents.java:
```diff
@@ -0,0 +1,66 @@
+	 * An event for the server play network handler receiving an update indicating the connected client's ability to receive packets in certain channels.
+	 *
+	 * @see PlayPacketSender#hasChannel(Identifier)
+	 */
```
Same here see below

Vaerian commented on fabric-networking-api-v1/src/main/java/net/fabricmc/fabric/impl/networking/AbstractChanneledNetworkAddon.java:
```diff
@@ -0,0 +1,173 @@
+import net.fabricmc.fabric.api.networking.v1.PacketByteBufs;
+import net.fabricmc.fabric.api.networking.v1.PlayPacketSender;
+
+// play
```
What's this comment for?

Vaerian commented on fabric-networking-v0/src/main/java/net/fabricmc/fabric/impl/networking/OldNetworkingHooks.java:
```diff
@@ -0,0 +1,36 @@
+import net.fabricmc.fabric.api.event.network.C2SPacketTypeCallback;
+import net.fabricmc.fabric.api.networking.v1.ServerChannelEvents;
+
+public final class OldNetworkingHooks implements ModInitializer {
```
This is definitely me missing the point, but what is this for exactly?

i509VCB replied:
That class implements the old `PacketType` events using the new `ClientChannelEvents` and `ServerChannelEvents`

Vaerian reviewed (approved):
Everything looks good I mainly just had some minor concerns over some of the documentation and beefing that up. After looking through it for a while it actually wasn't that bad making sense of all of it, but for the average user this whole module is kind of overwhelming so having more details about exactly when certain events are occurring and that sort of thing might be helpful. Other than that I saw a couple places where there were TODOs floating around, some of them looked like integrations with future modules, some just looked like things that have yet to be done. Regardless I'm sure they're in the works and will be done before final pull. To me it looks pretty good I don't really have any concerns. The one design detail I was hesitant about was having only a single handler per channel but that seems best for simplicity and people who want multiple handlers per channel just need to fix their impl because the only way I could see you wanting that would be from trying to implement a badly designed idea.

i509VCB committed:
Add null checks around the place, clarify some javadoc and method names

i509VCB commented on fabric-networking-api-v1/src/main/java/net/fabricmc/fabric/api/networking/v1/ChannelHandlerRegistry.java:
```diff
@@ -0,0 +1,51 @@
+	/**
+	 * Registers a handler to a channel.
+	 *
+	 * <p>If a handler is already registered to the {@code channel}, this method will return {@code false}, and no change will be made.
```
Imo specifying semantics are fine in javadoc for an interface

i509VCB committed:
Make FutureListeners uninstantiable

i509VCB committed:
Some internal nullable annotations

i509VCB committed:
An impl class I forgot to rename

i509VCB committed:
Some comments and clarify some client login handler javadoc

i509VCB committed:
Add a missing FunctionalInterface annotation

i509VCB committed:
Split play and login, move client stuff to right package

i509VCB committed:
No interface left behind

i509VCB committed:
Inline channel registries in api

i509VCB committed:
Login and play subpackages not needed

i509VCB committed:
Add helper method to create play custom packets

i509VCB committed:
hasGlobalChannel -> hasGlobalReceiver

i509VCB committed:
Just rename the collection method for now

i509VCB committed:
Inline PlayPacketSender into static methods

i509VCB committed:
Start on testmod idea for verifying dynamic registration

i509VCB committed:
Add client login events

i509VCB committed:
You don't say hello when talking to yourself.

Also more testmod stuff

i509VCB committed:
Make event names present tense

i509VCB committed:
Some javadoc and impl interface rename

i509VCB committed:
Change the test keybinding

i509VCB committed:
Begin working on dynamic reg

i509VCB committed:
Dynamic reg works, just need a lot of cleanup and reimpling global

i509VCB committed:
A few renames, readd global methods

i509VCB committed:
Try to reduce the amount of duplicate registration logic

i509VCB committed:
Reimplement dynamic accessors

i509VCB committed:
More impl

i509VCB committed:
Start reimplementing global receivers. Still very hacky solution.

i509VCB committed:
Reimplement some server global reciever stuff

i509VCB committed:
Add login init event for server login.

i509VCB committed:
Implement client login query start event

i509VCB committed:
Move event invocations into addon, don't dual register global recievers

i509VCB committed:
Finally reimplement global recievers for all networking phases

i509VCB committed:
A revelation: Send packets properly

This also finds the issue with screen getting the proper S2C channels, current on TODO list.

i509VCB committed:
Disconnect event does not need a packet sender

i509VCB committed:
Clarify, add methods to get channels net handler can recieve on client

i509VCB committed:
Unregister actually works now

i509VCB committed:
Bunch of null checks, add simpler login delay test for vanilla clients

i509VCB committed:
Add some debug logging entries, fix unregister on client's session reg

i509VCB committed:
Play channel event javadoc and rename login query handlers

i509VCB committed:
More channel -> channelName

i509VCB committed:
thisening

i509VCB committed:
Introduce the basics infrastructure for tracking global receivers

i509VCB committed:
Add more substantial javadoc to login connection events

i509VCB committed:
Javadoc, reimplement unreg methods on v0, 1 impl fix

i509VCB committed:
Implement tracking for global recievers

i509VCB committed:
Dont forget to start tracked sessions in 3/4 cases

i509VCB committed:
Global receiver docs and move methods in classes

i509VCB committed:
Complete null checks

i509VCB committed:
big boi javadoc part 1

i509VCB committed:
Finish the main javadoc, usage javadoc is left

i509VCB committed:
Set so has method is not needed

i509VCB committed:
Rename receiveable and sendable methods

i509VCB committed:
Add the two missing private ctors

i509VCB committed:
buildscript update to upstream

i509VCB committed:
Split out player finding stuff to networking player tracking API v1

Signed-off-by: liach <liach@users.noreply.github.com>

Forward v0 PlayerStream to new module, add entity track events

Rename module to player tracking

Well javadoc can make sense

Decide on tracking for the name

Update fabric-player-tracking-api-v1/src/main/java/net/fabricmc/fabric/api/networking/player/tracking/v1/package-info.java

Co-authored-by: Erlend Åmdal <erlend@aamdal.com>
Remove exceptions from javadoc that are not thrown

javadoc fix again

Handle a case where the player manager happens to be null

rename player tracking to player lookup

Yeet

i509VCB committed:
Cherrypick #1092

i509VCB committed:
Remove some redundant client networking methods, rename `(un)register` to `(un)registerReceiver`

i509VCB committed:
Simplify access to dynamic reg on client

i509VCB committed:
Param shifting, let users get sender.

i509VCB committed:
Warning about time and distance units

Player3324 reviewed (approved):
API looks good, thanks for this extensive effort!

shedaniel commented on fabric-networking-api-v1/src/main/java/net/fabricmc/fabric/api/client/networking/v1/ClientLoginConnectionEvents.java:
```diff
@@ -0,0 +1,99 @@
+/**
+ * Offers access to events related to the connection to a server on the client while the server is processing the client's login request.
+ */
+public final class ClientLoginConnectionEvents {
```
This (and its inner classes) should be annotated with @Environment(EnvType.CLIENT)

shedaniel commented on fabric-networking-api-v1/src/main/java/net/fabricmc/fabric/api/networking/v1/PlayerLookup.java:
```diff
@@ -0,0 +1,193 @@
+		return world(world)
+				.stream()
+				.filter((p) -> p.squaredDistanceTo(pos) <= radiusSq)
+				.collect(Collectors.toList());
```
Would returning an `Iterable` or `Iterator` be sufficient in order to not creating a list?

Player3324 replied:
Having the size is quite nice and it needs to create something, the (Array)List should be one of the cheapest options.

i509VCB committed:
Make sure these are client only

i509VCB committed:
Fix control flow in ClientPlayNetworking#send

i509VCB committed:
Correct example code javadoc

sschr15 reviewed (commented):
The copyright in the headers does not include 2020

haykam821 commented:
Don't worry, that isn't new: #598

i509VCB commented:
The license header year will be handled later when I make Fabric API use JDK style license headers

i509VCB committed:
javadoc correction in server login

modmuss50 commented on fabric-player-tracking-api-v1/src/main/resources/fabric.mod.json:
```diff
@@ -0,0 +1,26 @@
+{
```
I dont think this player tracking stuff belongs here? The module is not included in gradle?

modmuss50 commented on fabric-networking-api-v1/src/testmod/java/net/fabricmc/fabric/test/networking/login/NetworkingLoginQueryTest.java:
```diff
@@ -0,0 +1,76 @@
+				FutureTask<?> future = new FutureTask<>(() -> {
+					for (int i = 0; i <= 10; i++) {
+						Thread.sleep(300);
+						NetworkingTestmods.LOGGER.info("Delayed login for number {} 300 milliseconds", i);
```
Disable this by default, its a fine test just dont really want this all the time when using the test mods :D

modmuss50 commented on fabric-networking-api-v1/src/testmod/java/net/fabricmc/fabric/test/networking/keybindreciever/NetworkingKeybindPacketTest.java:
```diff
@@ -0,0 +1,54 @@
+
+	@Override
+	public void onInitialize() {
+		// FIXME: DO NOT LET ME PR THIS BEFORE FIXING IT
```
Need fixing apparenly.

modmuss50 reviewed (changes requested):
Looking really good 👍 few very minor things around the test mods.

This will need a decent amount of testing, including vanilla/mutliversion servers (where we have had issues before).

i509VCB committed:
Put login delay tests behind system property

Also remove unnecessary junk added by old module that was merged together.

i509VCB committed:
Fix ordering so channel registrations during `PHASE`_INIT work

i509VCB committed:
Fix prod bug and an oversight

i509VCB committed:
Fix login when connecting to dedicated server

i509VCB committed:
Update registry sync to v0 to prevent issue with reg sync hanging client

FlashyReese commented on fabric-networking-api-v1/src/main/java/net/fabricmc/fabric/impl/networking/server/QueryIdFactory.java:
```diff
@@ -0,0 +1,39 @@
+ */
+interface QueryIdFactory {
+	static QueryIdFactory create() {
+		// todo incremental ids or randomized
```
Is already done? Should remove

modmuss50 reviewed (approved):
Awesome 🍾 

Nice work, the API looks solid 👍

i509VCB committed:
this is done

NeusFear commented:
Sorry, had stashed comments apparently.

OroArmor reviewed (approved):
Ok its time to merge this in

modmuss50 commented:
Lets go 🚀 

Thanks for this.
